lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py

```diff
@@ -200,7 +200,10 @@ def _alloc_ip_address_from_dhcp(
             dhcp_desc = None
             while (
                 retry_count < self._retry_limit
-                and dhcp_allocated_ip(dhcp_desc) is not True
+                and (
+                    (dhcp_desc is None)
+                    ^ ((dhcp_desc is not None) and not dhcp_allocated_ip(dhcp_desc))
+                )
             ):
```

Comment:
I dont see reason to make this condition complicated? why is 'dhcp_desc' None check is moved from dhcp_allocated_ip() all call sites?

Reply:
This changes was triggered by a mypy issue which was tricky to remove. Moreover there was a inconsistency regarding null checks in the code, which was cleaned up with this refactoring. Those null checks are nowhere else moved to a deeper level but for this specific one. In addition having this check inside the method (combined with multiple levels of negation) made the logic harder comprehend in a lot if places. Now there is only one place, which has extended logic, which is also clearly visible and thus much more transparent. All other places usages benefit from a much clearer logical structure.